sycl/plugins/opencl/pi_opencl.hpp

```diff
@@ -17,11 +17,94 @@
 #ifndef PI_OPENCL_HPP
 #define PI_OPENCL_HPP
 
+#include <string>
+
 // This version should be incremented for any change made to this file or its
 // corresponding .cpp file.
 #define _PI_OPENCL_PLUGIN_VERSION 1
 
 #define _PI_OPENCL_PLUGIN_VERSION_STRING                                       \
   _PI_PLUGIN_VERSION_STRING(_PI_OPENCL_PLUGIN_VERSION)
 
+namespace OCLV {
+class OpenCLVersion {
+protected:
+  unsigned int major;
+  unsigned int minor;
+
+public:
+  OpenCLVersion() : major(0), minor(0) {}
+
+  OpenCLVersion(unsigned int major, unsigned int minor)
+      : major(major), minor(minor) {
+    if (!isValid())
+      major = minor = 0;
+  }
+
+  OpenCLVersion(const char *version) {
+    int ret;
+
+    /* The OpenCL specification defines the full version string as
+     * 'OpenCL<space><major_version.minor_version><space><platform-specific
+     * information>' for platforms and as
+     * 'OpenCL<space><major_version.minor_version><space><vendor-specific
+     * information>' for devices.
+     */
+    ret = sscanf(version, "OpenCL %u.%u", &major, &minor);
+    if (ret < 2 || !isValid())
+      major = minor = 0;
+  }
+
+  OpenCLVersion(const std::string &version) : OpenCLVersion(version.c_str()) {}
+
+  bool operator==(const OpenCLVersion &v) const {
+    return major == v.major && minor == v.minor;
+  }
+
+  bool operator!=(const OpenCLVersion &v) const { return !(*this == v); }
+
+  bool operator<(const OpenCLVersion &v) const {
+    if (major == v.major)
+      return minor < v.minor;
+    else
+      return major < v.major;
+  }
+
+  bool operator>(const OpenCLVersion &v) const { return v < *this; }
+
+  bool operator<=(const OpenCLVersion &v) const {
+    return (*this < v) || (*this == v);
+  }
+
+  bool operator>=(const OpenCLVersion &v) const {
+    return (*this > v) || (*this == v);
+  }
+
+  bool isValid() const {
```

Comment:
The idea of the function was to make sure that only versions of OpenCL which actually exist are accepted by ctors, whereas passing nonexistent versions results into creation of an object equal to OCLV::INVALID.

However, it's probably not that great if this function must be adjusted as soon as a newer OpenCL version is released, hence maybe it's just better to check for obviously bogus versions only.

Reply:
>> However, it's probably not that great if this function must be adjusted as soon as a newer OpenCL version is released, hence maybe it's just better to check for obviously bogus versions only.

Right, it would be unfortunate if a new OpenCL version is rejected just because of this function.
Assuming the new version supports everything previous version does.